### ISSUE TITLE
Frequent timeout failures in P2P statesync reactor test

### ISSUE THREAD

**Issue by creachadair** [tendermint/tendermint#7038]
In e2e test runs in CI, this test appears to time out fairly frequently:

```
go test -run 'TestReactor_StateProviderP2P' ./internal/statesync
```

@tychoish reports seeing a timeout at https://github.com/tendermint/tendermint/blob/1c9a6b105fe2aaae30a178e07bae6ee540b63281/internal/consensus/state_test.go#L344 with some frequency.

We aren't yet sure what causes the timeout, but suspect it may be blocking for synchronization with the event system. @williambanfield investigated some goroutine stack dumps and found a number of goroutines stalled in the pubsub package, possibly while trying to [`send`](https://github.com/tendermint/tendermint/blob/master/libs/pubsub/pubsub.go#L489-L527) events.

It's not clear where the culprit is, but since subscriptions are not closed or cleaned up, we suspect they could starve the pubsub service routine. 

We've made some initial attempts to corner the problem, cf. #7025. We'd like to understand what causes this failure and at least bound it before we cut a release for v0.35.

**Comment by creachadair** [comment]
The fixes from #7025 addressed at least some of the issues we identified, viz., possible deadlock while trying to update the write-ahead log.

**Comment by creachadair** [comment]
This failure hasn't cropped up for the past couple of days since #7025 merged, so I think we may be out of the woods here. Closing this for now on that basis.
